src/stream/src/executor/approx_percentile/local.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::iter;
+
+use risingwave_common::array::Op;
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+
+use crate::executor::prelude::*;
+
+pub struct LocalApproxPercentileExecutor {
+    _ctx: ActorContextRef,
+    pub input: Executor,
+    pub base: f64,
+    pub percentile_index: usize,
+    pub chunk_size: usize,
+}
+
+impl LocalApproxPercentileExecutor {
+    pub fn new(
+        _ctx: ActorContextRef,
+        input: Executor,
+        base: f64,
+        percentile_index: usize,
+        chunk_size: usize,
+    ) -> Self {
+        Self {
+            _ctx,
+            input,
+            base,
+            percentile_index,
+            chunk_size,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self) {
+        let percentile_index = self.percentile_index;
+        #[for_await]
+        for message in self.input.execute() {
+            match message? {
+                Message::Chunk(chunk) => {
+                    let mut builder = DataChunkBuilder::new(
+                        vec![DataType::Int16, DataType::Int32, DataType::Int32],
+                        self.chunk_size,
+                    );
+                    let chunk = chunk.project(&[percentile_index]);
+                    let mut pos_counts = HashMap::new();
+                    let mut neg_counts = HashMap::new();
+                    let mut zero_count = 0;
+                    for (op, row) in chunk.rows() {
+                        let value = row.datum_at(0).unwrap();
+                        let value: f64 = value.into_float64().into_inner();
+                        if value < 0.0 {
+                            let value = -value;
+                            let bucket = value.log(self.base).ceil() as i32; // TODO(kwannoel): should this be floor??
+                            let count = neg_counts.entry(bucket).or_insert(0);
+                            match op {
+                                Op::Insert | Op::UpdateInsert => *count += 1,
+                                Op::Delete | Op::UpdateDelete => *count -= 1,
```

Comment:
Is this able to handle the situation where `count` becomes negative? Considering that the local phase is really stateless, it should be able to provide an incremental value to the global phase.

Reply:
Count will never become negative. It will go to `0` at most.

Reply:
Why? Don't quite get it. If the local executor receives some `Delete`s, the local count can be negative IIUC.

Reply:
Oh sorry I misread. Yes for the local step, count can be negative. Not for the global step.